test/test_handlers.py

```diff
@@ -591,6 +590,91 @@ def guide():
     svi.update(svi_state)
 
 
+def test_collapse_normal_normal():
+    data = np.array(0.)
+
+    def model():
+        x = numpyro.sample("x", dist.Normal(0, 1))
+        with handlers.collapse():
+            y = numpyro.sample("y", dist.Normal(x, 1.))
+            numpyro.sample("z", dist.Normal(y, 1.), obs=data)
+
+    def guide():
+        loc = numpyro.param("loc", 0.)
+        scale = numpyro.param("scale", 1., constraint=constraints.positive)
+        numpyro.sample("x", dist.Normal(loc, scale))
+
+    svi = SVI(model, guide, numpyro.optim.Adam(1), Trace_ELBO())
+    svi_state = svi.init(random.PRNGKey(0))
+    svi.update(svi_state)
+
+
+def test_collapse_normal_normal_plate():
+    data = np.arange(5.)
+
+    def model():
+        x = numpyro.sample("x", dist.Normal(0, 1))
+        with handlers.collapse():
+            y = numpyro.sample("y", dist.Normal(x, 1.))
+            with handlers.plate("data", len(data)):
+                numpyro.sample("z", dist.Normal(y, 1.), obs=data)
+
+    def guide():
+        loc = numpyro.param("loc", 0.)
+        scale = numpyro.param("scale", 1., constraint=constraints.positive)
+        numpyro.sample("x", dist.Normal(loc, scale))
+
+    svi = SVI(model, guide, numpyro.optim.Adam(1), Trace_ELBO())
+    svi_state = svi.init(random.PRNGKey(0))
+    svi.update(svi_state)
+
+
+def test_collapse_normal_plate_normal():
+    data = np.arange(5.)
+
+    def model():
+        x = numpyro.sample("x", dist.Normal(0, 1))
+        with handlers.collapse():
+            with handlers.plate("data", len(data)):
+                y = numpyro.sample("y", dist.Normal(x, 1.))
+                numpyro.sample("z", dist.Normal(y, 1.), obs=data)
+
+    def guide():
+        loc = numpyro.param("loc", 0.)
+        scale = numpyro.param("scale", 1., constraint=constraints.positive)
+        numpyro.sample("x", dist.Normal(loc, scale))
+
+    svi = SVI(model, guide, numpyro.optim.Adam(1), Trace_ELBO())
+    svi_state = svi.init(random.PRNGKey(0))
+    svi.update(svi_state)
+
+
+def test_collapse_normal_mvn_mvn():
+    T, d, S = 5, 2, 3
+    data = jnp.ones((T, S))
+
+    def model():
+        x = numpyro.sample("x", dist.Exponential(1))
+        with handlers.collapse():
+            with numpyro.plate("d", d):
+                # TODO: verify that to_event works here
+                beta0 = numpyro.sample("beta0", dist.Normal(0, 1).expand([S]).to_event(1))
+                # TODO: address beta0 is a str, which cannot do infer_param_domain
+                beta = numpyro.sample("beta", dist.MultivariateNormal(beta0, jnp.eye(S)))
+            # FIXME: beta is a string here, how to apply numeric operators
+            mean = jnp.ones((T, d)) @ beta
```

Comment:
@fritzo @eb8680 looking like this is a good test to have better coverage for `collapse`. I think that we might return a Variable instead of a string, so we don't need to infer domain for `beta0` and here we can use `funsor.ops.matmul`.

Reply:
:+1: We fixed this to return a `Variable` in Pyro https://github.com/pyro-ppl/pyro/pull/2676

Regarding `funsor.ops.matmul` I would prefer the `@` version because it minimizes mention of magical funsors. I think that would work in `beta @ ones` but not in `ones @ beta`. I wonder if we can overload jnp.ones to return `NotImplemented` if faced with a Funsor so that `jnp.ones(...) @ ` would defer to `Funsor.__rmatmul__()`?

Reply:
Oops, thanks @fritzo! I forgot about it...

> wonder if we can overload `jnp.array.__matmul__()`

It seems that it works:
```python
import jax.numpy as jnp

class A():
    def __rmatmul__(self, other):
        return "a"

assert jnp.ones(3) @ A() == "a"
```

Reply:
Great, so `jnp.ones(...) @ funsor.Variable(...)` should also work!